src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -347,7 +365,33 @@ - (void)_handleEventReport:(NSArray<NSDictionary<NSString *, id> *> *)eventRepor
 {
     os_unfair_lock_lock(&self->_lock);
 
-    // first combine with previous unreported events, if they exist
+    // If event time is of MTREventTimeTypeSystemUpTime type, then update estimated start time as needed
+    NSDate * oldEstimatedStartTime = _estimatedStartTime;
+    for (NSDictionary<NSString *, id> * eventDict in eventReport) {
+        NSNumber * eventTimeTypeNumber = eventDict[MTREventTimeTypeKey];
+        if (!eventTimeTypeNumber) {
+            MTR_LOG_ERROR("Event %@ missing event time type", eventDict);
+            continue;
+        }
+        MTREventTimeType eventTimeType = (MTREventTimeType) eventTimeTypeNumber.unsignedIntegerValue;
+        if (eventTimeType == MTREventTimeTypeSystemUpTime) {
+            NSNumber * eventTimeValueNumber = eventDict[MTREventSystemUpTimeKey];
+            if (!eventTimeValueNumber) {
+                MTR_LOG_ERROR("Event %@ missing event time value", eventDict);
+                continue;
+            }
+            NSTimeInterval eventTimeValue = eventTimeValueNumber.doubleValue;
+            NSDate * potentialSystemStartTime = [NSDate dateWithTimeIntervalSinceNow:-eventTimeValue];
+            if (!_estimatedStartTime || ([potentialSystemStartTime compare:_estimatedStartTime] == NSOrderedAscending)) {
+                _estimatedStartTime = potentialSystemStartTime;
```

Comment:
So here's a question: if the device reboots, don't we need to reset _estimatedStartTime?

In particular, I think we should probably reset it if either of two things happens:

1. We lose subscription (because this may indicate a reboot).
2. We get a StartUp event (because the device rebooted but persisted subscriptions).

Or we need to update it even if it's going to increase, based on some heuristic about how we feel it's gone wrong.

Or something.

Reply:
Thank you for noting the StartUp event needing to reset this time, and I've made the change to do that.

I don't know what other heuristic I should employ here - if it's a "jumped forward quite a lot", I'm not sure how to convert that into code... And I'm not thinking of another way to tell if something's gone wrong.

Reply:
Well, the other heuristic I suggested is subscription loss.

Reply:
Done.